In [1]:
import pandas as pd
df = pd.read_csv(r"https://raw.githubusercontent.com/microbhai/CustomerChurnAnalysis/master/telecommunication_customer_churn.csv")

In [2]:
df.head(10)

CaseOrder Customer_id                           Interaction  \
0          1     K409198  aa90260b-4141-4a24-8e36-b04ce1f4f77b   
1          2     S120509  fb76459f-c047-4a9d-8af9-e0f7d4ac2524   
2          3     K191035  344d114c-3736-4be5-98f7-c72c281e2d35   
3          4      D90850  abfa2b40-2d43-4994-b15a-989b8c79e311   
4          5     K662701  68a861fd-0d20-4e51-a587-8a90407ee574   
5          6     W303516  2b451d12-6c2b-4cea-a295-ba1d6bced078   
6          7     U335188  6630d501-838c-4be4-a59c-6f58c814ed6a   
7          8     V538685  70ddaa89-b726-49dc-9022-2d655e4c7936   
8          9     M716771  05a49ee3-8fd5-453a-a5f3-82b6cd986856   
9         10     I676080  86f17e4d-2c24-4b70-a6ec-dddf0609dbaa   

                                UID           City State  \
0  e885b299883d4f9fb18e39c75155d990    Point Baker    AK   
1  f2de8bef964785f41a2959829830fb8a    West Branch    MI   
2  f1784cfa9f6d92ae816197eb175d3c71        Yamhill    OR   
3  dc8a365077241bb5cd5ccd305136b05e        Del Mar    CA   
4  aabb64a116e83fdc4befc1fbab1663f9      Needville    TX   
5  97598fd95658c80500546bc1dd312994    Fort Valley    GA   
6  87d1c4223e49156020564c01a88973b9        Pioneer    TN   
7  fce3f21888317907de42e298d718ccce  Oklahoma City    OK   
8  6c7043ced703b84de29766af3d53c976    Saint Cloud    FL   
9  6ffe183271258a039e122ced8750b2a0     Cincinnati    OH   

                  County    Zip       Lat        Lng  ...  OnlineBackup  \
0  Prince of Wales-Hyder  99927  56.25100 -133.37571  ...           Yes   
1                 Ogemaw  48661  44.32893  -84.24080  ...            No   
2                Yamhill  97148  45.35589 -123.24657  ...            No   
3              San Diego  92014  32.96687 -117.24798  ...            No   
4              Fort Bend  77461  29.38012  -95.80673  ...            No   
5                  Peach  31030  32.57032  -83.89040  ...           Yes   
6                  Scott  37847  36.43420  -84.27892  ...            No   
7               Oklahoma  73109  35.43313  -97.52463  ...           Yes   
8                Osceola  34771  28.27646  -81.16273  ...           Yes   
9               Hamilton  45237  39.19296  -84.45230  ...            No   

  DeviceProtection TechSupport StreamingTV  StreamingMovies  PaperlessBilling  \
0               No          No          No              Yes               Yes   
1               No          No         Yes              Yes               Yes   
2               No          No          No              Yes               Yes   
3               No          No         Yes               No               Yes   
4               No         Yes         Yes               No                No   
5              Yes          No          No              Yes                No   
6               No         Yes         Yes              Yes                No   
7               No          No          No               No               Yes   
8               No          No          No               No               Yes   
9              Yes          No          No              Yes               Yes   

              PaymentMethod     Tenure MonthlyCharge Bandwidth_GB_Year  
0   Credit Card (automatic)   6.795513    172.455519        904.536110  
1  Bank Transfer(automatic)   1.156681    242.632554        800.982766  
2   Credit Card (automatic)  15.754144    159.947583       2054.706961  
3              Mailed Check  17.087227    119.956840       2164.579412  
4              Mailed Check   1.670972    149.948316        271.493436  
5          Electronic Check   7.000994    185.007692       1039.357983  
6          Electronic Check  13.236774    200.118516       1907.242972  
7              Mailed Check   4.264255    114.950905        979.612708  
8  Bank Transfer(automatic)   8.220686    117.468591       1312.874964  
9              Mailed Check   3.422086    162.482694        508.763791  

[10 rows x 42 columns]

In [3]:
df.Churn.head(10)

0     No
1    Yes
2     No
3     No
4    Yes
5     No
6    Yes
7    Yes
8     No
9     No
Name: Churn, dtype: object

In [4]:
df.shape

(10000, 42)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 42 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   CaseOrder             10000 non-null  int64  
 1   Customer_id           10000 non-null  object 
 2   Interaction           10000 non-null  object 
 3   UID                   10000 non-null  object 
 4   City                  10000 non-null  object 
 5   State                 10000 non-null  object 
 6   County                10000 non-null  object 
 7   Zip                   10000 non-null  int64  
 8   Lat                   10000 non-null  float64
 9   Lng                   10000 non-null  float64
 10  Population            10000 non-null  int64  
 11  Area                  10000 non-null  object 
 12  TimeZone              10000 non-null  object 
 13  Job                   10000 non-null  object 
 14  Children              10000 non-null  int64  
 15  Age                 

In [6]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score


# Task 1: SIMPLE LOGISTIC REGRESSION

# The problem
### **Does monthly charges of our services likely to cause customer churn?**
We use binary logistic regression to establish whether a customer would churn out or not, and to what extent. 

The model would be: 


ln(p/1-p) = Bo+B1(Monthly charges) ==> p/1-p = e^(B0+B1(Monthly Charges))


Ho: B1 = 0 (The model is NOT statistically fit - The predictor is not a good predictor in predicting our response variable)

vs 

H1: B1 != 0 (The model is statistically fit. he predictor is a good predictor in predicting our response variable)

In [7]:
X = df.MonthlyCharge.values.reshape(-1, 1)
y = df['Churn'].apply(lambda x: 1 if x == 'Yes' else 0)

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
model = LogisticRegression(random_state=0).fit(X_train, y_train)
model.predict(X[:2, :])
model.predict_proba(X[:2, :])
model.score(X, y) #if >0.5, round off to 1 to imply the custmer is likely to churn

0.7569

In [9]:
import pandas as pd
import statsmodels.api as sm
from sklearn.model_selection import train_test_split

# Assuming df is your DataFrame
X = df['MonthlyCharge']
y = df['Churn'].apply(lambda x: 1 if x == 'Yes' else 0)

# Adding a constant to the model (intercept)
X = sm.add_constant(X)

# Splitting the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Fitting the logistic regression model using statsmodels
model = sm.Logit(y_train, X_train).fit()

# Printing the summary
print(model.summary())


Optimization terminated successfully.
         Current function value: 0.505264
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:                  Churn   No. Observations:                 8000
Model:                          Logit   Df Residuals:                     7998
Method:                           MLE   Df Model:                            1
Date:                Sat, 01 Jun 2024   Pseudo R-squ.:                  0.1235
Time:                        14:49:18   Log-Likelihood:                -4042.1
converged:                       True   LL-Null:                       -4611.4
Covariance Type:            nonrobust   LLR p-value:                1.320e-249
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
const            -4.8457      0.130    -37.313      0.000      -5.100      -4.591
MonthlyCharge     0.

ln(p/1-p) = -4.8457+0.0212(Monthly charges)

Since 1 is not included in the confidence interval for the value of the coefficient of Monthly charge, we reject Ho and conclude that Monthly charge has a significant effect on Customer Churn. Moreover, e^0.0212 = 1.0214; implying that a person is 102.14% more likely to churn out due to monthly charges.

# Multiple Logistic Regression

In [10]:
df.columns = ['CaseOrder', 'Customer_id', 'Interaction', 'UID', 'City', 'State',
       'County', 'Zip', 'Lat', 'Lng', 'Population', 'Area', 'TimeZone', 'Job',
       'Children', 'Age', 'Income', 'Marital', 'Gender', 'Churn',
       'Outage_sec_perweek', 'Email', 'Contacts', 'Yearly_equip_failure',
       'Techie', 'Contract', 'Port_modem', 'Tablet', 'InternetService',
       'Phone', 'Multiple', 'OnlineSecurity', 'OnlineBackup',
       'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
       'PaperlessBilling', 'PaymentMethod', 'Tenure', 'MonthlyCharge',
       'Bandwidth_GB_Year']

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 42 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   CaseOrder             10000 non-null  int64  
 1   Customer_id           10000 non-null  object 
 2   Interaction           10000 non-null  object 
 3   UID                   10000 non-null  object 
 4   City                  10000 non-null  object 
 5   State                 10000 non-null  object 
 6   County                10000 non-null  object 
 7   Zip                   10000 non-null  int64  
 8   Lat                   10000 non-null  float64
 9   Lng                   10000 non-null  float64
 10  Population            10000 non-null  int64  
 11  Area                  10000 non-null  object 
 12  TimeZone              10000 non-null  object 
 13  Job                   10000 non-null  object 
 14  Children              10000 non-null  int64  
 15  Age                 

In [12]:
df.Multiple.unique()

array(['No', 'Yes'], dtype=object)

In [13]:
df['OnlineBackup'].unique()

array(['Yes', 'No'], dtype=object)

In [14]:
df.columns

Index(['CaseOrder', 'Customer_id', 'Interaction', 'UID', 'City', 'State',
       'County', 'Zip', 'Lat', 'Lng', 'Population', 'Area', 'TimeZone', 'Job',
       'Children', 'Age', 'Income', 'Marital', 'Gender', 'Churn',
       'Outage_sec_perweek', 'Email', 'Contacts', 'Yearly_equip_failure',
       'Techie', 'Contract', 'Port_modem', 'Tablet', 'InternetService',
       'Phone', 'Multiple', 'OnlineSecurity', 'OnlineBackup',
       'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
       'PaperlessBilling', 'PaymentMethod', 'Tenure', 'MonthlyCharge',
       'Bandwidth_GB_Year'],
      dtype='object')

In [15]:
df.Techie.unique()

array(['No', 'Yes'], dtype=object)

In [20]:
df.Area.unique()

array(['Urban', 'Suburban', 'Rural'], dtype=object)

In [22]:
df_subset = df[['County', 'Area', 'Job', 'Marital', 'Gender', 'Techie', 'Contract', 'Port_modem', 'Tablet', 'InternetService', 'Phone', 'Multiple', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'PaperlessBilling', 'PaymentMethod','Children', 'Age', 'Income', 'Outage_sec_perweek', 'Yearly_equip_failure', 'MonthlyCharge', 'Bandwidth_GB_Year']]

                
df_subset = pd.get_dummies(df_subset, columns=['Techie'], prefix='Techie', drop_first=True)
df_subset = pd.get_dummies(df_subset, columns=['Area'], prefix='Area', drop_first=True)
df_subset = pd.get_dummies(df_subset, columns=['OnlineBackup'], prefix='OnlineBackup', drop_first=True)

dummy_columns = ['Techie_Yes', 'Area_Suburban' ,'OnlineBackup_Yes']
X = df_subset[dummy_columns].astype(int)
X.head(10)

Techie_Yes  Area_Suburban  OnlineBackup_Yes
0           0              0                 1
1           1              0                 0
2           1              0                 0
3           1              1                 0
4           0              1                 0
5           0              0                 1
6           1              1                 0
7           1              1                 1
8           0              1                 1
9           0              0                 0

In [19]:
df_subset.head(10)

County                                Job        Marital  \
0  Prince of Wales-Hyder  Environmental health practitioner        Widowed   
1                 Ogemaw             Programmer, multimedia        Married   
2                Yamhill            Chief Financial Officer        Widowed   
3              San Diego                          Solicitor        Married   
4              Fort Bend                Medical illustrator      Separated   
5                  Peach           Chief Technology Officer  Never Married   
6                  Scott             Surveyor, hydrographic        Widowed   
7               Oklahoma  Sales promotion account executive        Married   
8                Osceola     Teaching laboratory technician      Separated   
9               Hamilton           Museum education officer        Married   

      Gender        Contract Port_modem Tablet InternetService Phone Multiple  \
0       Male        One year        Yes    Yes     Fiber Optic   Yes       No   
1     Female  Month-to-month         No    Yes     Fiber Optic   Yes      Yes   
2     Female        Two Year        Yes     No             DSL   Yes      Yes   
3       Male        Two Year         No     No             DSL   Yes       No   
4       Male  Month-to-month        Yes     No     Fiber Optic    No       No   
5     Female        One year        Yes     No             NaN   Yes      Yes   
6       Male  Month-to-month         No     No             DSL   Yes       No   
7     Female  Month-to-month         No     No             DSL    No       No   
8  Nonbinary  Month-to-month        Yes     No             DSL   Yes       No   
9     Female        Two Year        Yes     No     Fiber Optic   Yes       No   

   ... Yearly_equip_failure MonthlyCharge Bandwidth_GB_Year Techie_No  \
0  ...                    1    172.455519        904.536110      True   
1  ...                    1    242.632554        800.982766     False   
2  ...                    1    159.947583       2054.706961     False   
3  ...                    0    119.956840       2164.579412     False   
4  ...                    1    149.948316        271.493436      True   
5  ...                    1    185.007692       1039.357983      True   
6  ...                    1    200.118516       1907.242972     False   
7  ...                    0    114.950905        979.612708     False   
8  ...                    3    117.468591       1312.874964      True   
9  ...                    0    162.482694        508.763791      True   

  Techie_Yes Area_Rural Area_Suburban  Area_Urban  OnlineBackup_No  \
0      False      False         False        True            False   
1       True      False         False        True             True   
2       True      False         False        True             True   
3       True      False          True       False             True   
4      False      False          True       False             True   
5      False      False         False        True            False   
6       True      False          True       False             True   
7       True      False          True       False            False   
8      False      False          True       False            False   
9      False       True         False       False             True   

   OnlineBackup_Yes  
0              True  
1             False  
2             False  
3             False  
4             False  
5              True  
6             False  
7              True  
8              True  
9             False  

[10 rows x 31 columns]

In [17]:
df.iloc[11,11]

'Rural'

In [23]:
X = X
y = df['Churn'].apply(lambda x: 1 if x == 'Yes' else 0)

# Adding a constant to the mode-l (intercept)
X = sm.add_constant(X)

# Splitting the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Fitting the logistic regression model using statsmodels
model = sm.Logit(y_train, X_train).fit()

# Printing the summary
print(model.summary())

Optimization terminated successfully.
         Current function value: 0.572755
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:                  Churn   No. Observations:                 8000
Model:                          Logit   Df Residuals:                     7996
Method:                           MLE   Df Model:                            3
Date:                Sat, 01 Jun 2024   Pseudo R-squ.:                0.006373
Time:                        14:56:15   Log-Likelihood:                -4582.0
converged:                       True   LL-Null:                       -4611.4
Covariance Type:            nonrobust   LLR p-value:                 1.072e-12
                       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------
const               -1.2088      0.042    -28.924      0.000      -1.291      -1.127
Techie_Yes 